### PR TITLE
API: Add .mT attribute for arrays

### PR DESCRIPTION
**Pull request by Kai-Striega**
This is the initial commit for adding an `mT`, or matrix transpose, attribute for arrays. It is still a work in progress and not done yet. Please note that this is my first time working on NumPy's C-API. The PR is mainly being submitted this early to confirm that I am on the right track.

Things that are still needed:

- [x] Release note
- [x] Masked array support
- ~[ ] `np.transpose`~
- [x] Check input dimensions in `array_matrix_transpose_get` and exit early if 1d array
- [x] Tests fail for higher dimensional arrays
- [x] Figure out the correct location for the tests

Another note; `ndarray.T` has a corresponding function `np.transpose`, as part of this PR should we add a corresponding function for `.mT` e.g. `np.matrix_transpose`? If this is the case, should I add a function to `PyArray_MatrixTranspose` to `shape.c` rather than defining the function inside `getset.c` as is done with the regular transpose function?

### CODE REVIEW

File: numpy/core/tests/test_matrix_transpose.py
Comment (Kai-Striega) [inline numpy/core/tests/test_matrix_transpose.py:23]: This is pretty naive. It could create a very large array, that may cause problems depending on which shapes are generated.

File: numpy/core/tests/test_matrix_transpose.py
Comment (Kai-Striega) [inline numpy/core/tests/test_matrix_transpose.py:9]: Here and below the test name needs to be renamed to use transpose, not swapaxes.

File: numpy/core/src/multiarray/getset.c
Comment (seberg) [reply numpy/core/src/multiarray/getset.c:938]: Even if it raises, I expect we should have an explicit error anyway for clarity.

File: numpy/core/tests/test_matrix_transpose.py
Comment (seberg) [reply numpy/core/tests/test_matrix_transpose.py:1]: There is probably no great place... Could start an `test_arrayobject.py` if you don't find a nice place elsewhere.

File: numpy/core/tests/test_matrix_transpose.py
Comment (seberg) [reply numpy/core/tests/test_matrix_transpose.py:23]: Use `broadcast_to`?  Trusts the content is correct, but that seems OK (at least beyond 1-2 explicit tests).  TBH, hypothesis seems a bit overkill in this case, its not like there is any weird complexity we need to cover beyond high dimensions.

**Review by seberg — COMMENTED** [review]
Looks like a good start, will mark as draft for now if you don't mind, just unmark it when things are settling (not that I think there is much missing)!

File: numpy/core/src/multiarray/getset.c
Comment (Kai-Striega) [reply numpy/core/src/multiarray/getset.c:938]: I'm not sure if an explicit error is the way to go. Without any changes, `.mT` returns the same as `.T` which is what we want, right? I was thinking along the lines of making it explicit (if `ndim == 1` return `self`).

File: numpy/core/tests/test_matrix_transpose.py
Comment (Kai-Striega) [reply numpy/core/tests/test_matrix_transpose.py:25]: This test now passes. Turns out CLion doesn't rebuild every time I run the test suite. I was using an outdated build.

File: numpy/core/tests/test_matrix_transpose.py
Comment (Kai-Striega) [reply numpy/core/tests/test_matrix_transpose.py]: Done.

File: numpy/core/tests/test_matrix_transpose.py
Comment (Kai-Striega) [reply numpy/core/tests/test_matrix_transpose.py]: Done.

File: numpy/core/src/multiarray/getset.c
Comment (seberg) [reply numpy/core/src/multiarray/getset.c:938]: no, it has to error `ndim < 2`, otherwise it gives the wrong idea in formulas.  Maybe adding a dimension might work, but that would require a lot of convincing, so I think it's a no-go right now, to be clear, I am not convinced yet.  (And I don't actually think it is that clear whether someone wants a column or row vector even if we have an implicit decision anyway.)

Comment (Kai-Striega) [reply numpy/core/src/multiarray/getset.c:938]: Ok. I'll follow your advice. Would a `ValueError` be a good fit here? I've had a quick look through the available errors and couldn't find anything that fits better.

Comment (seberg) [reply numpy/core/src/multiarray/getset.c:944]: I think that is what we use for shape related errors always, so that sounds good to me.  We were discussing introducing a shape or broadcast error, but that is a different project.

**Comment by rgommers** [comment]
Thanks for working on this Kai!

> Another note; `ndarray.T` has a corresponding function `np.transpose`, as part of this PR should we add a corresponding function for `.mT` e.g. `np.matrix_transpose`?

I'd say yes we want that (because of the analogue to `.T` and because [the array API standard](https://data-apis.org/array-api/latest/API_specification/generated/array_api.matrix_transpose.html) has it). It doesn't have to be in this PR though, if it's already large and easier to do as a follow-up - whatever seems best to you.

File: numpy/core/code_generators/numpy_api.py
Comment (Kai-Striega) [inline numpy/core/code_generators/numpy_api.py:373]: I've done some reading on what `MinVersion` does. It's defined in numpy/core/code_generators/genapi.py

File: numpy/core/src/multiarray/methods.c
Comment (Kai-Striega) [inline numpy/core/src/multiarray/methods.c:2405]: Does this need to be explicitly cast to `PyObject *`

**Comment by Kai-Striega** [comment]
@rgommers I'm happy to work on it in this PR, but I'm still learning about NumPy internals, meaning that the commit history could become quite messy as I learn (of course I'll try my best to keep it clean). As a summary, I think we should match the functions for `transpose` i.e.:

* `np.matrix_transpose`
* `np.ndarray.mT`
* `np.ndaray.matrix_transpose`
* `np.ma.matrix_transpose`
* `np.ma.MaskedArray.mT`
* `np.ma.MaskedArray.matrix_transpose`

The next step for me is to add `np.ndaray.matrix_transpose`, which I am having some trouble with. What I have done so far is:

* Moved the implementation into a new function (`PyArray_MatrixTranspose`) into `numpy/core/src/multiarray/shape.c` alongside `PyArray_Transpose` 
* Added a `matrix_transpose` entry to `array_methods` in `numpy/core/src/multiarray/methods.c`
* Added `PyArray_Transpose` to the `multiarray_funcs_api` in `numpy/core/code_generators/numpy_api.py`

I'm not really sure if this is the right direction, or if we even want to add `PyArray_MatrixTranspose` to the C-API as (I think) I am doing.

**Comment by mattip** [comment]
I would not touch the NumPy C-API with this PR. It is enough to add the attribute to ndarray and ma. If in the future we wish to export it via the C-API, that can be another PR.

**Comment by rgommers** [comment]
> As a summary, I think we should match the functions for `transpose` i.e.:

We don't need/want both a `.mT` property and a `.matrix_transpose` method. It's duplicate and the plan is to trim down on `ndarray` methods. Just the property and the `np.matrix_transpose` function (and their MaskedArray equivalents) is enough.

**Comment by charris** [comment]
I rather like the idea of an `.mT` property :) It would be useful to have a short way to call it in `...@...@...` constructions.

**Comment by Kai-Striega** [comment]
I've removed the code adding it to the C-API (I think) and stopped working on adding an `ndarray.matrix_transpose` method. I've kept the function `PyArray_MatrixTranspose` in `shape.c` as I want to be able to share it between `ndarray.mT` and `np.matrix_transpose` - this is causing the test to segfault compared to the earlier version. 

I may need some help structuring the code so that I can share it between `ndarray.mT` and `np.matrix_transpose`.

File: numpy/core/code_generators/numpy_api.py
Comment (seberg) [reply numpy/core/code_generators/numpy_api.py]: That is correct, but I am not sure I feel we need to add this?  At some point people aimed to have a full featured C-API, I at least don't really feel there is a point for many convenience functions in C.

File: numpy/core/src/multiarray/methods.c
Comment (seberg) [reply numpy/core/src/multiarray/methods.c]: No, since `PyArray_MatrixTranspose` returns `PyObject *`, so it matches.  Otherwise you would see a compiler warning anyway.

File: numpy/core/code_generators/numpy_api.py
Comment (seberg) [reply numpy/core/code_generators/numpy_api.py]: MinVersion says that this is the first version that defines the function.

Comment (Kai-Striega) [reply numpy/core/code_generators/numpy_api.py]: > but I am not sure I feel we need to add this? 

I've decided it's not worth adding. My idea was to copy `np.transpose` so that I have something to go off. That isn't the way the reviewers want me to go, I'll leave it out.

> MinVersion says that this is the first version that defines the function.

Thanks

**Comment by seberg** [comment]
You have to add the new function to a header file now and include it, probably.  If you check the test failures, you see they fail on compiler warnings.

**Review by BvB93 — COMMENTED** [review]
Can you add `ndarray.mT` annotations to the main __init__.pyi stub file? The type is identical to `ndarray.T` so you can just reuse that it:
https://github.com/numpy/numpy/blob/cc0abd768575d7f9e862de0b4912af27f6e9690d/numpy/__init__.pyi#L986-L987

**Comment by Kai-Striega** [comment]
@seberg I've *finally* got the masked array implementation working! I'm not sure if it's done correctly, could you take a look at my last commit? 

If it's right, I think I have three things left:

- [ ] Take the reference to the function versions out of the release notes
- [ ] Add the documentation
- [ ] Add type hints for the masked array

File: numpy/ma/core.py
Comment (seberg) [inline numpy/ma/core.py:6108]: A masked array.  Otherwise, seems good.  Although `.transpose()` has some `_update_from_self` stuff, it may make sense to just stick close to it in the off-chance it matters.

Comment (Kai-Striega) [reply numpy/ma/core.py:6108]: I'm struggling to replicate `.transpose()` as it uses the function version, not the attribute. I haven't been figured out how to do this (yet) and will leave it for another PR - maybe I can come back and tidy it up then if there are problems?

**Comment by Kai-Striega** [comment]
@rgommers @seberg I've made the last changes that (I think) are needed, could you please ask you two for another review?

I've decided to leave the `np.matrix_transpose` function for now, I'll get to it in another PR

File: numpy/core/_add_newdocs.py
Comment (seberg) [inline numpy/core/_add_newdocs.py:2936]:
```suggestion

    .. versionadded:: 2.0

```

File: numpy/ma/core.py
Comment (seberg) [inline numpy/ma/core.py:6104]:
```suggestion

    .. versionadded:: 2.0

```

**Review by seberg — APPROVED** [review]
LGTM, thanks, just should add the versionadded.

**Comment by seberg** [comment]
Thanks, lets put it in.

After merging, I noticed that the docs actually render slightly incorrectly.  Not a big deal, but maybe you want to follow up?

I guess we need `<BLANKLINE>` or maybe its fine to just remove the blank line also.

**Comment by Kai-Striega** [comment]
Thanks that's my first C level contribution 👌. Let's hope for many more.

I'll follow up with the docs rendering properly.

**Pull request merged**
